langlib/langlib-test/src/test/resources/test-src/statements/foreach/foreach-on-fail.bal

```diff
@@ -93,3 +93,412 @@ function testNestedForeachLoopBreak() returns string {
     }
     return result;
 }
+
+type SampleErrorData record {|
+    int code;
+    string reason;
+|};
+
+type SampleError error<SampleErrorData>;
+
+string testMessage = "";
+int testErrorCode = 0;
+string testErrorReason = "";
+
+function testSimpleOnFailWithErrorBP() {
+    foreach var i in data {
+        fail error("error!");
+    } on fail error error(msg) {
+        testMessage = msg;
+    }
+     assertEquality(testMessage, "error!");
+}
+
+function testSimpleOnFailWithErrorBPWithVar() {
+    foreach var i in data {
+        fail error("error!");
+    } on fail var error(msg) {
+        testMessage = msg;
+    }
+     assertEquality(testMessage, "error!");
+}
+
+function testOnFailWithErrorBPHavingUserDefinedTypeWithError() {
+    foreach var i in data {
+        fail error SampleError("error!", code = 20, reason = "deadlock condition");
+    } on fail error error(msg, code = errCode, reason = errReason) {
+        testErrorCode = errCode;
+        testErrorReason = errReason;
+        testMessage = msg;
+    }
+    assertEquality(testMessage, "error!");
+    assertEquality(testErrorCode, 20);
+    assertEquality(testErrorReason, "deadlock condition");
+    assertEquality(testErrorCode is int, true);
+    assertEquality(testErrorReason is string, true);
```

Comment:
I wouldn't use module-level variables, won't allow us to properly run these tests concurrently.

How about asserting withing the on fail block and returning there. Then outside we can have something like an assert fail to make sure control was transferred to the on fail block.